### PR TITLE
feat(analyze): add modal for adding collaborators to analyze page

### PR DESCRIPTION
**Pull request by rohitkrai03**
Planner Story - https://openshift.io/openshiftio/Openshift_io/plan/detail/846
Invision Link - https://redhat.invisionapp.com/share/MVOBQ1ZRXHZ#/screens/322783163

- Adds modal for adding collaborators directly from analyze page.

Demo

### CODE REVIEW

File: src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts [hunk 1]
```diff
@@ -98,4 +107,23 @@ export class EditSpaceDescriptionWidgetComponent implements OnInit {
     return this.contexts.current.pipe(map(val => val.user.id === this.loggedInUser.id));
   }
 
+  launchAddCollaborators() {
+    this.modalAdd.show();
+  }
+
+  addCollaboratorsToParent(addedUsers: User[]) {
```
Comment (aptmac) [inline src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts]: same as above

File: src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts [hunk 2]
```diff
@@ -98,4 +107,23 @@ export class EditSpaceDescriptionWidgetComponent implements OnInit {
...
+    addedUsers.forEach(user => {
+      let matchingUser = find(this.collaborators, (existing) => {
```
Comment (aptmac) [inline src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts]: missing types

File: src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts [hunk 3]
```diff
@@ -98,4 +107,23 @@ export class EditSpaceDescriptionWidgetComponent implements OnInit {
...
+        return existing.id === user.id;
+      });
+      if (!matchingUser) {
+        this.collaborators.push(user);
+      }
+    });
+  }
+
+  ngOnDestroy() {
```
Comment (aptmac) [inline src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts]: same as above

File: src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts [hunk 4]
```diff
@@ -37,16 +41,20 @@ export class EditSpaceDescriptionWidgetComponent implements OnInit {
   ) { }
 
   ngOnInit() {
-    this.userService.loggedInUser.subscribe(val => this.loggedInUser = val);
-    this.spaces.current
+    this.subscriptions.push(this.userService.loggedInUser.subscribe(val => this.loggedInUser = val));
```
Comment (aptmac) [inline src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts]: missing types

File: src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts [hunk 5]
```diff
@@ -37,16 +41,20 @@ export class EditSpaceDescriptionWidgetComponent implements OnInit {
...
+    this.subscriptions.push(this.spaces.current
       .subscribe(space => {
         this.space = space;
         if (space) {
-          this.collaboratorCount = this.collaboratorService.getInitialBySpaceId(space.id).pipe(map(c => c.length));
+          this.subscriptions.push(
+            this.collaboratorService.getInitialBySpaceId(space.id).subscribe(users => {
```
Comment (aptmac) [inline src/app/dashboard-widgets/edit-space-description-widget/edit-space-description-widget.component.ts]: missing types

**Comment by rohitkrai03** [comment]
@aptmac Added types and sorting based on username.

File: src/app/space/settings/collaborators/add-collaborators-dialog/add-collaborators-dialog.component.ts [hunk 6]
```diff
@@ -37,15 +37,16 @@ export class AddCollaboratorsDialogComponent implements OnInit {
...
       this.collaborators = users;
+      this.sortCollaborators();
       this.loading = false;
     }, () => {
       this.collaborators = [];
     });
   }
 
-  public onOpen() {
+  onOpen() {
```
Comment (aptmac) [inline src/app/space/settings/collaborators/add-collaborators-dialog/add-collaborators-dialog.component.ts]: missing `: void`

**Comment by christianvogt** [comment]
@rohitkrai03 I spoke with @serenamarie125 about sorting and she suggests ascending sort on last name, followed by first name, followed by username. Please update accordingly.

**Comment by serenamarie125** [comment]
Please order the items in the add collaborators dialog to have the following sort order:
- primary - last name ascending
- secondary - first name ascending
- tertiary - username ascending

**Comment by rohitkrai03** [comment]
@serenamarie125 @christianvogt I do not understand the three levels of sorting (primary, secondary and tertiary) here. Also, we do not have user data saved as `first name` and `last name`. We just have `fullName` and `username`.

**Comment by serenamarie125** [comment]
@rohitkrai03 if you have 2 user ids with the same name ... it would be shown like this:

Serena Nichols (serenamarie125)
Serena Nichols (snichols@redhat.com)

**Comment by rohitkrai03** [comment]
This is basically what will be shown if the list is sorted based on username.

> Serena Nichols (serenamarie125)
> Serena Nichols (snichols@redhat.com)

There is a problem of not having first name and last name as well.
> Also, we do not have user data saved as first name and last name. We just have fullName and username.

**Comment by rohitkrai03** [comment]
@serenamarie125 Implemented two levels of sorting as discussed. 

- Primary - `fullName`
- Secondary - `username`

**Comment by serenamarie125** [comment]
great thank you @rohitkrai03 !

**Comment by joshuawilson** [comment]
For now, we can sort this way. But if the design and desire is to sort by Last name then we can either figure it out in the UI or get the API changed.

**Pull request merged**